libyara/modules/pe/pe.c

```diff
@@ -1067,9 +1099,278 @@ static IMPORTED_DLL* pe_parse_imports(PE* pe)
+    //  DelayImportNameTableRva         0x004010c0
+    //  BoundDelayImportTableRva        0x00000000
+    //  ...
+
+    if (virtual_address > image_base)
+    {
+      virtual_address = virtual_address - image_base;
+    }
+  }
+
+  return virtual_address;
+}
+
+
+int pe_is_termination_delay_import_entry(PIMAGE_DELAYLOAD_DESCRIPTOR importDescriptor)
+{
+  return (importDescriptor->Attributes.AllAttributes == 0 &&
+          importDescriptor->DllNameRVA == 0 &&
+          importDescriptor->ModuleHandleRVA == 0 &&
+          importDescriptor->ImportAddressTableRVA == 0 &&
+          importDescriptor->ImportNameTableRVA == 0 &&
+          importDescriptor->BoundImportAddressTableRVA == 0 &&
+          importDescriptor->UnloadInformationTableRVA == 0 &&
+          importDescriptor->TimeDateStamp == 0);
+}
+
+
+char * pe_parse_delay_import_dll_name(PE* pe, uint64_t rva)
+{
+  const uint64_t offset = pe_rva_to_offset(pe, rva);
+
+  if (offset < 0)
+    return NULL;
+
+  char* dll_name = (char *) (pe->data + offset);
+  if (!pe_valid_dll_name(dll_name, pe->data_size - (size_t) offset))
+    return NULL;
+  return yr_strdup(dll_name);
+}
+
+
+uint64_t pe_parse_delay_import_pointer(PE* pe, uint64_t pointerSize, uint64_t rva)
+{
+  const uint64_t offset = pe_rva_to_offset(pe, rva);
+  const uint8_t *data = pe->data + offset;
+
+
+  if (!fits_in_pe(pe, data, pointerSize))
+    return YR_UNDEFINED;
+
+
+  if (IS_64BITS_PE(pe))
+    return yr_le64toh(*(uint64_t *) data);
+  else
+    return yr_le32toh(*(uint32_t *) data);
+}
+
+static void* pe_parse_delayed_imports(PE* pe)
+{
+  int64_t offset;
+  uint64_t num_imports = 0;           // Number of imported DLLs
+  uint64_t num_function_imports = 0;  // Total number of functions imported
+  uint64_t image_base = OptionalHeader(pe, ImageBase);
+  uint64_t size_of_image = OptionalHeader(pe, SizeOfImage);
+  uint64_t pointer_size = (IS_64BITS_PE(pe)) ? 8 : 4;
+  uint64_t ordinal_mask = (IS_64BITS_PE(pe)) ? IMAGE_ORDINAL_FLAG64 : IMAGE_ORDINAL_FLAG32;
+
+  IMPORTED_DLL* head_dll = NULL;
+  IMPORTED_DLL* tail_dll = NULL;
+
+  IMPORT_FUNCTION *head_fun = NULL;
+  IMPORT_FUNCTION *tail_fun = NULL;
+
+  PIMAGE_DELAYLOAD_DESCRIPTOR import_descriptor = NULL;
+  PIMAGE_DATA_DIRECTORY directory = NULL;
+
+  // Default to 0 imports until we know there are any
+  set_integer(0, pe->object, "number_of_delayed_imports");
+  set_integer(0, pe->object, "number_of_delayed_imported_functions");
+
+
+  directory = pe_get_directory_entry(
+      pe, IMAGE_DIRECTORY_ENTRY_DELAY_IMPORT);
+
+  if (directory == NULL)
+    return NULL;
+
+  if (yr_le32toh(directory->VirtualAddress) == 0)
+    return NULL;
+
+
+  offset = pe_rva_to_offset(pe, yr_le32toh(directory->VirtualAddress));
+
+  if (offset < 0)
+    return NULL;
+
+  import_descriptor = (PIMAGE_DELAYLOAD_DESCRIPTOR)(pe->data + offset);
+  for (; struct_fits_in_pe(pe, import_descriptor, IMAGE_DELAYLOAD_DESCRIPTOR); import_descriptor++)
```

Comment:
It looks like this line is past the 80 columns limit used in all the code base. I encourage using `clang-format` for auto-formatting the code.

Reply:
Formatted using `clang-format`.